File: t2p_model.py
import t2p_funciones as t2p
import t2p_graph as graph


def results(parameters):
    T1 = [0, 0]
    T2 = [0, 0]
    solution = {
        "CFa": str(0),
        "Pa": str(0),
        "CFb": str(0),
        "Pb": str(0),
        "qa": str(0),
        "qb": str(0),
        "Q": str(0),
        "ECa": str(0),
        "ECb": str(0),
        "CT": str(0),
        "CF": str(0),
        "CV": str(0),
        "Cme": str(0),
        "beneficios": str(0),
        "beneficios_a": str(0),
        "beneficios_b": str(0),
        "RPa": False,
        "RPb": False,
        "RCIa": False,
        "RCIb": False,
    }
    if parameters == 0:
        return solution
    elif (
        (parameters["constant_1"] == "")
        or (parameters["constant_2"] == "")
        or (parameters["slope_1"] == "")
        or (parameters["slope_2"] == "")
    ):
        return solution
    else:
        T1[0] = float(parameters["constant_1"])
        T1[1] = float(parameters["slope_1"])
        T2[0] = float(parameters["constant_2"])
        T2[1] = float(parameters["slope_2"])
        proporcion = float(parameters["proporcion"])
        numero_clientes = int(parameters["numero_clientes"])
        cf = float(parameters["costo_fijo"])
        cmg = float(parameters["costo_marginal"])

        # Tarifas
        Pa = t2p.p_a(cmg)
        Pb = t2p.p_b(T1, T2, proporcion, cmg)
        CFa = t2p.cf_a(T1, T2, Pa, Pb)
        CFb = t2p.cf_b(T2, Pb)
        # Cantidades demandadas
        qa = t2p.q(T1, Pa)
        qb = t2p.q(T2, Pb)
        Q = numero_clientes * (proporcion * qa + (1 - proporcion) * qb)
        # Costos
        CT = cf + cmg * Q
        CV = cmg * Q
        CMe = CT / Q
        # Excedentes del conumidor
        ECa = t2p.ec(T1, Pa)
        ECb = t2p.ec(T2, Pb)
        # Beneficios
        beneficios = t2p.beneficios(
            numero_clientes, proporcion, T1, T2, Pa, Pb, cmg, cf
        )
        beneficios_a = (qa * (Pa - cmg) + CFa) * proporcion * numero_clientes
        beneficios_b = (qb * (Pb - cmg) + CFb) * (1 - proporcion) * numero_clientes

        # Restricción de participación

        if ECa < CFa:
            RPa = True
        else:
            RPa = False

        if ECb < CFb:
            RPb = True
        else:
            RPb = False

        # Restricción de compatibilidad de incentivos

        if (ECa - CFa) < (t2p.ec(T1, Pb) - CFb):
            RCIa = True
        else:
            RCIa = False

        if (ECb - CFb) < (t2p.ec(T2, Pa) - CFa):
            RCIb = True
        else:
            RCIb = False

        solution = {
            "CFa": str(CFa),
            "Pa": str(Pa),
            "CFb": str(CFb),
            "Pb": str(Pb),
            "qa": str(qa),
            "qb": str(qb),
            "Q": str(Q),
            "CT": str(CT),
            "CF": str(cf),
            "CV": str(CV),
            "Cme": str(CMe),
            "ECa": str(ECa),
            "ECb": str(ECb),
            "beneficios": str(beneficios),
            "beneficios_a": str(beneficios_a),
            "beneficios_b": str(beneficios_b),
            "RPa": RPa,
            "RPb": RPb,
            "RCIa": RCIa,
            "RCIb": RCIb,
        }
        return solution


def graphics(parameters):
    T1 = [0, 0]
    T2 = [0, 0]
    T1[0] = float(parameters["constant_1"])
    T1[1] = float(parameters["slope_1"])
    T2[0] = float(parameters["constant_2"])
    T2[1] = float(parameters["slope_2"])
    prop = float(parameters["proporcion"])
    CMg = float(parameters["costo_marginal"])
    return graph.drawgraph(T1, T2, CMg, prop)

File: t2p_model.js
import * as t2p from "./t2p_functions.js"

D1 = [0, 0]
D2 = [0, 0]

function t2pSolucion(parametros) {
    let solucion = {
        cargoFijoA : 0,
        precioA : 0,
        cargoFijoB : 0,
        precioB : 0,
        cantidadA : 0,
        cantidadB : 0,
        cantidadTotal : 0,
        excedenteConsumidorA: 0,
        excedenteConsumidorB: 0,
        costoTotal: 0,
        costosFijos: 0,
        costosVariables: 0,
        costoMedio: 0,
        beneficiosTotales : 0,
        beneficiosConsumidoresA: 0,
        beneficiosConsumidoresB: 0,
        restriccionParticipacionA: false,
        restriccionParticipacionB: false,
        restriccionCompIncentivosA: false,
        restriccionCompIncentivosB: false,
    }

    if (parametros !== 0) {
        // Guardamos los parámetros necesarios para el cálculo de los resultados
        DA[0] = parametros.constanteA
        DA[1] = parametros.pendienteA
        DB[0] = parametros.constanteB
        DB[1] = parametros.pendienteB
        propA = parametros.proporcionA
        numeroClientes = parametros.numeroClientes
        cf = parametros.costoFijo
        cmg = parametros.costoMarginal

        // Cálculo de tarifas
        pA = t2p.precioA(cmg)
        solucion.precioA = pA

        pB = t2p.precioB(DA, DB, propA, cmg)
        solucion.precioB = pB

        cargoFijoA = t2p.cfA(DA, DB, pA, pB)
        solucion.cargoFijoA = cargoFijoA

        cargoFijoB = t2p.cfB(DB, pB)
        solucion.cargoFijoB = cargoFijoB

        // cantidades demandadas
        qa = t2p.cantDem(DA, pA)
        solucion.cantidadA = qa
        qb = t2p.cantDem(DB, pB)
        solucion.cantidadB = qb
        Q = numeroClientes * (porpA * qa - (1 - propA) * qb)
        solucion.cantidadTotal = Q

        // Costos
        CT = cf + cmg * Q
        solucion.costoTotal = CT

        solucion.costosFijos = cf

        CV = cmg * Q
        solucion.costosVariables = CV

        CMe = CT / Q
        solucion.costoMedio = CMe

        // Excedentes del consumidor
        ecA = t2p.excedente(DA, pA)
        solucion.excedenteConsumidorA = ecA
        ecB = t2p.excedente(DB, pB)
        solucion.excedenteConsumidorB = ecB

        // Beneficios
        solucion.beneficiosTotales = t2p.beneficiosTotales(numeroClientes, propA, DA, DB, pA, pB, cmg, cf)
        
        // Restricción de compatibilidad de incentivos

        if (ecA >= cargoFijoA) {
            solucion.restriccionParticipacionA = true
        }

        if (ecB >= cargoFijoB) {
            solucion.restriccionParticipacionB = true
        }

        // Restricción de compatibilidad de incentivos
        if ((ecA - cargoFijoA) >= (t2p.excedente(DA,pB)-cargoFijoB)) {
            solucion.restriccionCompIncentivosA = true
        }
        if ((ecB - cargoFijoB) >= (t2p.excedente(DB,pA)-cargoFijoA)) {
            solucion.restriccionCompIncentivosB = true
        }
    }
    
    return solucion
}



